tapchannel/aux_funding_controller.go

```diff
@@ -51,6 +52,28 @@ const (
 	// ackTimeout is the amount of time we'll wait to receive the protocol
 	// level ACK from the remote party before timing out.
 	ackTimeout = time.Second * 30
+
+	// maxNumAssetIDs is the maximum number of fungible asset pieces (asset
+	// IDs) that can be committed to a single channel. The number needs to
+	// be limited to prevent the number of required HTLC signatures to be
+	// too large for a single CommitSig wire message to carry them. This
+	// value is tightly coupled with the number of HTLCs that can be added
+	// to a channel at the same time (maxNumHTLCs). The values were
+	// determined with the TestMaxCommitSigMsgSize test in
+	// aux_leaf_signer_test.go then a set was chosen that would allow for
+	// a decent number of HTLCs (and also a number that is divisible by two
+	// because each side will only be allowed to add half of the total).
+	maxNumAssetIDs = 3
```

Comment:
Why do we need to limit this, instead of just the total amount of HTLCs?

Reply:
In that if we have 10 passive asset IDs in a channel, there're no issues as long as the total amount of HTLCs isn't exceeded.

Reply:
Given a certain number of asset IDs would a proof-size limit be encountered?

Reply:
IIUC no, the real limit is the HTLCs. This PR seems to take an approach to optimize equal HTLC utilization across all asset IDs, but I think we can just limit HLTCs instead. 

If we move the TAP level HTLC sigs to a new message, then we don't need a limit at all. As we'll fall back to the existing limit at the base LN level.

Reply:
This conversation was resolved, but I agree with what Roas says above. Why are we moving forward with the tap level HTLC limitation in this PR?

Reply:
I resolved the comment based on this: https://github.com/lightninglabs/taproot-assets/pull/1132#issuecomment-2399339907

Tried my best to explain why this is needed in the code comments, but let me try again:
 - By our definition, within an on-chain transaction of multiple different involved asset IDs (e.g. multiple tranches of the same grouped asset), each asset ID needs to get its own virtual transaction.
 - A virtual transaction can combine multiple inputs of the _same_ asset ID into a single output. But if there are multiple input from different asset IDs (and because of that multiple virtual transactions), that also means there are multiple asset outputs in such a transaction.
 - If multiple different asset IDs are committed to a TAP channel, even if the on-chain funding output is one _logical_ shared UTXO, there would be an asset-level UTXO for each asset ID.
 - Depending on the amounts of those asset-level UTXOs and the size of the HTLC in question, each HTLC in a channel commitment transaction would spend between one and all of these UTXOs, therefore potentially requiring multiple asset-level signatures per HTLC.
 - We can't predict how many "pieces" of the channel inputs are consumed by each HTLC, so we have to make some sort of worst-case assumption. That worst case assumption is that each HTLC would need to carry an asset-level signature for each of the spent asset ID "piece". This is obviously an incorrect assumption as the worst-case can't really ever be true. But because we can't easily come up with a heuristic that always works, we might as well use the worst case and draw our conclusions from there.
 - All Lightning p2p messages are limited to 65k payload.
 - The question with everything above now is: How many worst-case asset HTLCs can we fit into a single [`commitment_signed`](https://github.com/lightning/bolts/blob/master/02-peer-protocol.md#the-commitment_signed-message) message? That's what the test answers.
 - Because there currently is no way to reject accepting an HTLC when noticing it would be too large at run time, we have to agree on a maximum number of in-flight HTLCs upfront, which means at channel open time.

I hope the above makes sense. Let me know if you think this would be helpful as a standalone document somewhere (not sure where would make sense).

But, keep in mind: This is kind of a short-term stop-gap solution, as mentioned below. The "real" solution would be to circumvent the p2p message payload size limit by allowing a single message to be split up into multiple smaller ones (which requires a protocol update).

Reply:
Thanks for the detailed response. Very valuable.

> But, keep in mind: This is kind of a short-term stop-gap solution, as mentioned below. The "real" solution would be to circumvent the p2p message payload size limit by allowing a single message to be split up into multiple smaller ones (which requires a protocol update).

Seems like a complex solution for a temporary problem. Does it make more sense to limit naively (without care for worst case asset utxo count) for now and work on the split message side of things?

Does this really require a protocol update? Can't we just use our own message types that support splitting?

Reply:
> Does it make more sense to limit naively

What values would you suggest? Sure, for now we could go with just 1 asset ID since we can't yet commit more than one anyway. But as soon as we add grouped asset support, this _will_ become an issue.

And remember, picking the wrong value means force close as worst case scenario, which is definitely not what we want.

> Can't we just use our own tap message types that support splitting?

Sure, since we do lnd-to-lnd only at the moment, we can do it without a protocol update. But that's still quite a big refactor on the `lnd` side. So not sure we can wait for that.

Reply:
So we _can_ do a protocol upgrade, but those changes would be more invasive. Right now we piggy back on top of all the existing LN related wire messages, so we can rely on `lnd` to do serialization and also retransmission for us. The state machine will also handle all messages, and pass over context to `tapd` as needed. 

If we add a new message, then we need to make sure that we process this in lock step with the existing LN channel state machine. Eg: if we add `asset_htlc_sigs`, we need to make sure that we validate that fully _before_ we revoke a state, or we can end up with an invalid commitment. 

I think this is fine as a stopgap, as the 483 limit is relatively high in practice. Most other implementations have a global limit of this value to like 50 or so. We can monitor to see if this is too low in practice, but it's also the case that right now with the way the protocol works, anyone can offer us a smaller limit. 

I think the path we can pursue after we land this is: revive the old message fragmentation logic in lnd. This keeps the 65 KB protocol limit, but lets an application level message span across multiple logical transport messages. With that, we can lift these limits, and things work as normal for the most part. We'd just need to make sure that re-introducing that logic doesn't break things for normal LN, in particular cross impl compat.